### PR TITLE
Spatial filtering of far fields by windowing near fields

### PR DESCRIPTION
**Pull request by shashwat-sh**
Backend PR: https://github.com/flexcompute/tidy3d-core/pull/468

Addresses this issue: https://github.com/flexcompute/tidy3d-core/issues/395

Notebook that demonstrates this: https://github.com/flexcompute-readthedocs/tidy3d-docs/pull/329

### CODE REVIEW

File: tidy3d/components/monitor.py [hunk 1]
```diff
@@ -666,6 +667,46 @@ class AbstractFieldProjectionMonitor(SurfaceIntegrationMonitor, FreqMonitor):
...
+        (0, 0),
+        title="Spatial filtering window size",
+        description="Size of the transition region of the windowing function used to ensure that "
+        "the recorded near fields decay to zero near the edges of the monitor. "
+        "The two components refer to the two tangential directions associated with each surface. "
+        "For surfaces with the normal along ``x``, the two components are (``y``, ``z``). "
+        "For surfaces with the normal along ``y``, the two components are (``x``, ``z``). "
+        "For surfaces with the normal along ``z``, the two components are (``x``, ``y``). "
+        "Each value must be between 0 and 1, inclusive, and denotes the size of the transition "
+        "region over which fields transition to near-0, relative to half the size of the monitor "
```
Comment (tylerflex) [inline tidy3d/components/monitor.py:680]: How is near-0 defined?

**Review by tylerflex — COMMENTED** [review]
Thanks @shash-sharma just a few comments. otherwise looks good.

File: tidy3d/components/field_projection.py [hunk 2]
```diff
@@ -459,6 +459,41 @@ def integrate_for_one_theta(i_th: int):
...
+        """Apply windowing function to the surface currents."""
+        if proj_monitor.size.count(0.0) == 0:
+            return currents
+        if proj_monitor.window_size == (0, 0):
+            return currents
+
+        pts = [currents[name].values for name in ["x", "y", "z"]]
+
+        custom_bounds = [
+            [pts[i][0] for i in [0, 1, 2]],
```
[the remark this answers is not in the document]
Comment (shashwat-sh) [reply tidy3d/components/field_projection.py:475]: I think this would only work if each of the arrays in `pts` is of equal size?

**Review by momchil-flex — APPROVED** [review]

File: tidy3d/components/field_projection.py [hunk 3]
```diff
@@ -596,17 +630,20 @@ def _project_fields_cartesian(
 
             for surface in self.surfaces:
```
[the remark this answers is not in the document]
Comment (tylerflex) [reply tidy3d/components/field_projection.py:643]: yea not a high priority just something I noticed.

File: tidy3d/components/monitor.py [hunk 4]
```diff
@@ -666,6 +670,47 @@ class AbstractFieldProjectionMonitor(SurfaceIntegrationMonitor, FreqMonitor):
...
+    window_size: Tuple[pydantic.NonNegativeFloat, pydantic.NonNegativeFloat] = pydantic.Field(
+        (0, 0),
+        title="Spatial filtering window size",
+        description="Size of the transition region of the windowing function used to ensure that "
+        "the recorded near fields decay to zero near the edges of the monitor. "
+        "The two components refer to the two tangential directions associated with each surface. "
+        "For surfaces with the normal along ``x``, the two components are (``y``, ``z``). "
+        "For surfaces with the normal along ``y``, the two components are (``x``, ``z``). "
+        "For surfaces with the normal along ``z``, the two components are (``x``, ``y``). "
+        "Each value must be between 0 and 1, inclusive, and denotes the size of the transition "
```
[the remark this answers is not in the document]
Comment (shashwat-sh) [reply tidy3d/components/monitor.py:682]: Lumerical defines its `alpha` w.r.t. _half_ the size of the monitor, so a value of 1 corresponds to a transition region spanning half the monitor on one side, and half on the other, and I followed that convention. I think I mentioned the "half" thing in the docstring but can emphasize / clarify further what a value of 1 means

**Pull request merged**